### ISSUE TITLE
Provide instances for OneAnd

### ISSUE THREAD

**Issue by travisbrown** [travisbrown/circe#170]
From @mpilquist on [Gitter](https://gitter.im/travisbrown/circe?at=56a24a39dc33b33c754747e2). We currently provide encoders and decoders for `NonEmptyList`, but these instances could be for any `OneAnd` where there's an instance for the collection type.

**Comment by ArtemPyanykh** [comment]
I wonder what you mean by "_where there's an instance for the collection type_"? 

As far as I understand even if we have an `Encoder` for `F[A]` in case of `OneAnd` we need to be able to put an additional value `head: A` inside a context `tail: F[A]` before applying the encoder. If we're speaking only about `NonEmptyList/Vector/Stream` it might be enough to have:
```
def encodeOneAnd[A: Encoder, F[_]: Foldable](x: OneAnd[F, A]): Encoder[OneAnd[F, A]]
```

For decoder this won't work though. Basically we can always decode `JArray` into `List`. So to implement `Decoder[OneAnd[F, A]]` we need to transform `List[A]` into `F[A]`. It seems to me that the minimum needed set of capabilities on `F` is having `empty`, `pure` and `combine`. `MonadCombine` has all these and defined for `List`, `Vector` and `Stream` which gives out of box support for all `NonEmpty*` from `cats`.

Another things is this reasoning is applicable only to _collection-like_ containers, and decoding from/into `JArray`. It's not clear what should be a generic representation for, say, `OneAnd[Id, A]` which is isomorphic to `(A, A)`.

Or am I thinking in the wrong direction?

**Comment by travisbrown** [comment]
@ArtemPyanykh You're right that there are a few ways this could be done, and I think it's fine to only target collection-like `OneAnd`s.

One other approach would be to require only instances for both `A` and `F[A]` for the `OneAnd[F, A]` codecs. For the encoder you apply the `F[A]` encoder to the tail and then prepend the encoded head _if_ the `F[A]` encoder returned a `JArray` (and fail otherwise). For the decoder you first check whether the cursor points to a non-empty `JArray`. If it doesn't, you fail, but if it does you decode the first `Json` with the `A` decoder and the rest with the `F[A]` decoder.

This approach takes advantage of the optimized `CanBuildFrom` decoder for all the `NonEmpty*` types in cats.

What do you think?

**Comment by ArtemPyanykh** [comment]
@travisbrown not sure about this part "prepend the encoded head if the F[A] encoder returned a JArray (and fail otherwise)". The signature of Encoder's apply `def apply(a: A): Json` assumes we should always be able to encode `OneAnd[F, A]`, so I'm not sure at which point we could gracefully fail.

As for the decoder -- I need to think a bit. Basically my concern is `OneAnd[List, Int]` would be encoded like `[1, 2, 3, 4]` which is OK and natural. 
However it seems to me that to decode it into `OneAnd[List, Int]` _reusing_ existing list decoder we would need to pass it something like `[1, [2, 3, 4]]` instead of `[1, 2, 3, 4]` since decoder for `CanBuildFrom` would try after extracting `head = 1` to
1. do `downArray` on an `HCursor` pointing at `[1, -->2, 3, 4]` which would fail,
2. check if it's an array under the focus, which would also fail since `c.focus` would return `JNumber(2)`.

I'm not that familiar with the library, though, so I can be mistaken in this regard.

**Comment by travisbrown** [comment]
@ArtemPyanykh Oh, you're right about encoding (sorry—in the middle of something else and not paying enough attention). Then `Foldable[F]` probably makes sense for `Encoder`.

Here's a quick sketch of the idea for decoding:

```scala
Decoder.instance { c =>
  val beginning = c.downArray
  for {
    h <- beginning.as[Int]
    t <- beginning.right.deleteLefts.up.as[List[Int]]
  } yield (h, t)
}
```

It might be possible to improve the implementation but that basic idea should work.

**Comment by ArtemPyanykh** [comment]
@travisbrown ah, didn't think we could just remove the first element of an array :smile: Cool, let me try that one then.

**Comment by travisbrown** [comment]
Fixed by #178. Thanks again, @ArtemPyanykh.

**Comment by ArtemPyanykh** [comment]
@travisbrown no problem, looking forward to contribute more. This `beginner-friendly` flag is really helpful :+1:
